File: backend/app/Domain/Customers/Models/Customer.php
<?php

declare(strict_types=1);

namespace App\Domain\Customers\Models;

use App\Domain\Pricing\Models\PriceType;
use App\Domain\Warehouses\Models\Warehouse;
use App\Models\User;
use Database\Factories\CustomerFactory;
use Illuminate\Database\Eloquent\Factories\Factory;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\SoftDeletes;

/**
 * @property int $id
 * @property string $code
 * @property string $name
 * @property bool $is_company
 * @property string|null $contact_name
 * @property string|null $phone
 * @property string|null $email
 * @property string|null $address
 * @property string|null $city
 * @property string|null $ice
 * @property int|null $price_type_id
 * @property int|null $seller_id
 * @property int|null $warehouse_id
 * @property float $credit_limit
 * @property float $balance
 * @property bool $is_blocked
 * @property string|null $notes
 * @property bool $is_active
 */
class Customer extends Model
{
    /** @use HasFactory<CustomerFactory> */
    use HasFactory, SoftDeletes;

    protected $fillable = [
        'code',
        'name',
        'is_company',
        'contact_name',
        'phone',
        'email',
        'address',
        'city',
        'ice',
        'price_type_id',
        'seller_id',
        'warehouse_id',
        'credit_limit',
        'balance',
        'is_blocked',
        'notes',
        'is_active',
        'created_by',
    ];

    protected function casts(): array
    {
        return [
            'is_company' => 'boolean',
            'credit_limit' => 'decimal:2',
            'balance' => 'decimal:2',
            'is_blocked' => 'boolean',
            'is_active' => 'boolean',
        ];
    }

    /**
     * @return Factory<self>
     */
    protected static function newFactory(): Factory
    {
        return CustomerFactory::new();
    }

    /**
     * Crédit disponible = plafond − encours.
     */
    public function availableCredit(): float
    {
        return round((float) $this->credit_limit - (float) $this->balance, 2);
    }

    /**
     * Type de prix appliqué par défaut (détail / demi-gros / gros).
     *
     * @return BelongsTo<PriceType, $this>
     */
    public function priceType(): BelongsTo
    {
        return $this->belongsTo(PriceType::class);
    }

    /**
     * Vendeur référent.
     *
     * @return BelongsTo<User, $this>
     */
    public function seller(): BelongsTo
    {
        return $this->belongsTo(User::class, 'seller_id');
    }

    /**
     * Lieu de rattachement.
     *
     * @return BelongsTo<Warehouse, $this>
     */
    public function warehouse(): BelongsTo
    {
        return $this->belongsTo(Warehouse::class);
    }

    /**
     * Utilisateur ayant créé le client (portée de visibilité).
     *
     * @return BelongsTo<User, $this>
     */
    public function createdBy(): BelongsTo
    {
        return $this->belongsTo(User::class, 'created_by');
    }
}

File: mobile/lib/models/customer.dart
/// Client (CustomerResource).
class Customer {
  const Customer({
    required this.id,
    required this.code,
    required this.name,
    this.phone,
    this.city,
    this.email,
    this.address,
    required this.creditLimit,
    required this.balance,
    required this.availableCredit,
    required this.isBlocked,
    required this.isActive,
  });

  final int id;
  final String code;
  final String name;
  final String? phone;
  final String? city;
  final String? email;
  final String? address;
  final double creditLimit;

  /// Encours (solde dû par le client).
  final double balance;
  final double availableCredit;
  final bool isBlocked;
  final bool isActive;

  bool get isOverLimit => creditLimit > 0 && balance > creditLimit;

  factory Customer.fromJson(Map<String, dynamic> json) => Customer(
        id: json['id'] as int,
        code: json['code'] as String? ?? '',
        name: json['name'] as String? ?? '',
        phone: json['phone'] as String?,
        city: json['city'] as String?,
        email: json['email'] as String?,
        address: json['address'] as String?,
        creditLimit: (json['credit_limit'] as num?)?.toDouble() ?? 0,
        balance: (json['balance'] as num?)?.toDouble() ?? 0,
        availableCredit: (json['available_credit'] as num?)?.toDouble() ?? 0,
        isBlocked: json['is_blocked'] == true,
        isActive: json['is_active'] != false,
      );
}

/// Écriture du relevé client (GET /customers/{id}/statement).
class StatementEntry {
  const StatementEntry({
    this.date,
    required this.type,
    required this.amount,
    required this.balanceAfter,
    this.note,
  });

  final String? date;
  final String type;
  final double amount;
  final double balanceAfter;
  final String? note;

  factory StatementEntry.fromJson(Map<String, dynamic> json) => StatementEntry(
        date: json['date'] as String?,
        type: json['type'] as String? ?? '',
        amount: (json['amount'] as num?)?.toDouble() ?? 0,
        balanceAfter: (json['balance_after'] as num?)?.toDouble() ?? 0,
        note: json['note'] as String?,
      );
}
